ext/nnstreamer/tensor_filter/tensor_filter_tvm.cc

```diff
@@ -0,0 +1,413 @@
+ * @brief   NNStreamer tensor-filter sub-plugin for Apache TVM
+ * @author  Junhwan Kim <jejudo.kim@samsung.com>
+ * @see     http://github.com/nnstreamer/nnstreamer
+ * @bug     No known bugs
+ * 
+ * This is the per-NN-framework plugin (TVM) for tensor_filter.
+ * 
+ * @note    Only supports tvm.contrib.graph_executor model
+ */
+
+#include <glib.h>
+#include <nnstreamer_cppplugin_api_filter.hh>
+#include <nnstreamer_log.h>
+#include <tensor_common.h>
+
+#include <dlpack/dlpack.h>
+#include <tvm/runtime/module.h>
+#include <tvm/runtime/packed_func.h>
+#include <tvm/runtime/registry.h>
+
+
+namespace nnstreamer
+{
+namespace tensorfilter_tvm
+{
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+void init_filter_tvm (void) __attribute__ ((constructor));
+void fini_filter_tvm (void) __attribute__ ((destructor));
+#ifdef __cplusplus
+}
+#endif /* __cplusplus */
+
+
+/**
+ * @brief Tensor data information to execute the model
+ */
+typedef struct {
+  std::vector<int64_t> shape;
+  DLDataType dtype;
+} tvm_data_info;
+
+/**
+ * @brief Class for TVM subplugin.
+ */
+class tvm_subplugin final : public tensor_filter_subplugin
+{
+  private:
+  bool empty_model;
+  char *model_path;
+  GstTensorsInfo inputInfo;
+  GstTensorsInfo outputInfo;
+
+  DLDevice device;
+  tvm::runtime::Module mod_factory;
+  tvm::runtime::Module gmod;
+  std::vector<tvm_data_info> input_info_list;
+  std::vector<tvm_data_info> output_info_list;
+
+  static const char *name;
+  static const accl_hw hw_list[];
+  static tvm_subplugin *registeredRepresentation;
+
+  bool parse_custom_prop (const char *custom_prop);
+  bool convert_nns_type (tensor_type nns_type, DLDataType &tvm_type);
+  void cleanup();
+  bool configure_meta (GstTensorsInfo &tensor_meta, std::vector<tvm_data_info> &tensor_list);
+  bool set_tensor_info (GstTensorsInfo &dest_info, const GstTensorsInfo &src_info, unsigned int num_tensors);
+
+  public:
+  static void init_filter_tvm ();
+  static void fini_filter_tvm ();
+
+  tvm_subplugin ();
+  ~tvm_subplugin ();
+
+  tensor_filter_subplugin &getEmptyInstance ();
+  void configure_instance (const GstTensorFilterProperties *prop);
+  void invoke (const GstTensorMemory *input, GstTensorMemory *output);
+  void getFrameworkInfo (GstTensorFilterFrameworkInfo &info);
+  int getModelInfo (model_info_ops ops, GstTensorsInfo &in_info, GstTensorsInfo &out_info);
+  int eventHandler (event_ops ops, GstTensorFilterFrameworkEventData &data);
+};
+
+const char *tvm_subplugin::name = "tvm";
+const accl_hw tvm_subplugin::hw_list[] = { ACCL_CPU, ACCL_GPU };
+
+/**
+ * @brief Construct a new tvm subplugin::tvm subplugin object
+ */
+tvm_subplugin::tvm_subplugin ()
+  : tensor_filter_subplugin (), empty_model (true), model_path (nullptr),
+    device (DLDevice{kDLCPU, 0}), mod_factory (nullptr), gmod (nullptr)
+{
+  inputInfo.num_tensors = 0;
+  outputInfo.num_tensors = 0;
+}
+
+/**
+ * @brief Cleanup method for tvm subplugin
+ */
+void tvm_subplugin::cleanup ()
```

Comment:
I'd recommend cleanup function to guarantee `noexcept` (and check if it is no except) as this is being used in the destructor

Reply:
useful keyword! thx